hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/FSHLog.java

```diff
@@ -412,6 +422,24 @@ protected void doReplaceWriter(Path oldPath, Path newPath, Writer nextWriter) th
     }
   }
 
+  private void closeWriter(Path path, boolean syncCloseCall) throws IOException {
+    try {
+      TraceUtil.addTimelineAnnotation("closing writer");
+      writer.close();
+      TraceUtil.addTimelineAnnotation("writer closed");
+    } catch (IOException ioe) {
+      int errors = closeErrorCount.incrementAndGet();
+      boolean hasUnflushedEntries = isUnflushedEntries();
+      if (syncCloseCall && (hasUnflushedEntries || (errors > this.closeErrorsTolerated))) {
```

Comment:
confirming once again with hasFlushedEntries when syncCloseCall is it mandatory or just to be on the safe side? Because even if hasUnFlushedEntries is false here i think errors would have crossed the limit anyway because we increment and then check it. 
Rest LGTM. Can check the QA once. May be you may have to change some cluster based test to run with FSHLog as the default may be AsyncWAL.

Reply:
Ideally when we writer close, there wont be any unflushed entries. We have synced all writes happend to this wal writer.  But as the code was having it and to be on safe side, continue to use that.  Ya in that case will go with sync way of call so that we can act against and IOE in close call.  Same with that errors count. If consecutively those many failures, we will have to throw IOE and RS will abort.  But here one catch is there.  This #errors based abort wont be that strict from now on.  Say if the errors tolerated is 1, we need to throw back IOE if errors count is 2.  But it is possible that one close req came and given to a thread and before that is done another close req and given to another thread and then a 3rd close req came by then 1st thread is done and it was failed. So the 3rd req will be tried in sync way of close. By the time it finishes, the 2nd close thread also failed to close. So the errors already became 2 and 3d  (sync way call) also failed which makes the count to 3.  So instead of throw back IOE at errors=2, it will do now on errors=3.  But we are sure all edits are synced and no data loss. So I believe this is ok.. Asycn wal is not having any such logic at all.

Reply:
That is right. It is not a harm to check again. Am just saying that since the first sync based close was checking for >= it means either it has already reached the max count or already greater. And next time when we do throw the IOE we check it  after increment. Reg AsyncWAL i just said if tests are running with AsyncWAL this code path may never get executed. So should you go and change the WAL type for test cases to execute this flow?